Add MovePrivateIP and its OpenStack implementation

dulek opened the pull request:
Up until now when an IP is failed over from one node to another, `CloudPrivateIPConfigController` will call `ReleasePrivateIP` with the node being failed over and `AssignPrivateIP` with the new node.

This is far from ideal for OpenStack platform. In this case in addition to adding the IP to the `allowedAddressPairs` on the node's port, we also create a port in Neutron to act as a reservation of that IP, so that it won't be taken by another node being created. With traditional model on every failover the reservation port would get removed, making a race condition "stealing" the IP possible. It's also problematic if user assigned a Neutron floating IP to that reservation port to have a predictible SNAT address for EgressIP traffic going outside the cluster. On failover such association would be lost and manual intervention would be needed to assign FIP to new reservation port.

This commit introduces an optional `CloudProviderWithMoveIntf` interface that has `MovePrivateIP` operation to be called on IP fail over. This operation is supposed to move the IP assignment from one node to another, but doesn't require deleting cloud resources if not necessary. The required logic is added to `CloudPrivateIPConfigController` to call `MovePrivateIP` on IP fail over if plugin implements the `CloudProviderWithMoveIntf` interface.

An implementation of `MovePrivateIP` is added to the OpenStack plugin, that will remove the IP from old node's `allowedAddressPairs`, add the IP to the new node's `allowedAddressPairs`, but won't touch the Neutron port that serves as a reservation of that IP.

mandre commented on pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go:
```diff
@@ -179,24 +179,98 @@ func (c *CloudPrivateIPConfigController) SyncHandler(key string) error {
 
-	// At most one of nodeToAdd or nodeToDel will be set
-	nodeToAdd, nodeToDel := c.computeOp(cloudPrivateIPConfig)
+	// At most one of nodeNameToAdd or nodeNameToDel will be set
```
No longer true with your change. Please update the comments.

dulek replied:
I removed it and added a detailed comment to computeOp.

mandre reviewed (commented):
Mostly comments about comments :)
The code looks good to me. I'm not quite sure what causes [the unit tests failure](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cloud-network-config-controller/95/pull-ci-openshift-cloud-network-config-controller-master-unit/1615389218903691264) but it doesn't look related to this patch as it failed in a similar fashion [for two other PRs recently](https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-openshift-cloud-network-config-controller-master-unit).

andreaskaris commented on pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go:
```diff
@@ -179,24 +179,98 @@ func (c *CloudPrivateIPConfigController) SyncHandler(key string) error {
+		// This is a blocking call. If the IP is not assigned then don't treat
+		// it as an error.
+		var i interface{} = c.cloudProviderClient
+		withMover, ok := i.(cloudprovider.CloudProviderWithMoveIntf)
```
nit: I find the below way easier to read, but that may be personal preference:
~~~
withMover, ok := c.cloudProviderClient.(cloudprovider.CloudProviderWithMoveIntf)
~~~

andreaskaris commented on pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go:
```diff
@@ -179,24 +179,98 @@ func (c *CloudPrivateIPConfigController) SyncHandler(key string) error {
+		if !ok {
+			return fmt.Errorf("cannot convert driver to the interface with move abilities, this should never happen")
+		}
+		if moveErr := withMover.MovePrivateIP(ip, nodeToAdd, nodeToDel); moveErr != nil && !errors.Is(moveErr, cloudprovider.NonExistingIPError) {
```
extreme nit (also for other lines): I like keeping lines at 120 chars max + a few if I really cannot avoid it (which may be the case here). I know that's not the case for the older code in the repo and we don't really have a standard, but it's easier to read IMO. Feel free to ignore this and keep as is

andreaskaris replied:
Looking at the code, I think ReleasePrivateIP is the only function atm that can throw the NonExistingIPError (which is logical, in case of a Release, we have a means of reporting the issue, but we ignore it). I'm not sure if it makes sense for MovePrivateIP to throw such an error. Currently as far as I see it doesn't, and so we don't have to deal with it hear, either?

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	// Loop over all ports that are attached to this nova instance.
+	for _, serverPort := range serverPorts {
+		if isIPAddressAllowedOnNeutronPort(serverPort, ip) {
+			if err = o.unallowIPAddressOnNeutronPort(serverPort.ID, ip); err != nil {
```
Does the comment in 332 refer to this line here? If so, I'd say we need to return the error, because we check if the IP is in the allowed_address_pairs, and only if that's the case, we try to remove it (and retry). So if this still fails, then something went wrong during the detach / update of allowed_address_pairs. Returning the error will make sure that we do not continue without removing the IP from the allowed_address_pair on the source port. This is also important because without this succeeding, findAssignSubnetAndPort might return AlreadyExistingIPError

dulek replied:
We do remove the error. The comment is about - what if for each of the ports `isIPAddressAllowedOnNeutronPort()` was false?

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
 	return fmt.Errorf("could not assign IP address %s to node %s", ip, node.Name)
 }
 
+func (o *OpenStack) MovePrivateIP(ip net.IP, nodeToAdd, nodeToDel *corev1.Node) error {
```
I think what's missing is that we must account for the crazy corner case where the reservation port is gone (perhaps an admin deleted it) and must be recreated by MovePrivateIP?

dulek replied:
Then we should probably take care of a crazy corner case when the reservation port exists in AssignPrivateIP too. I'll look into this.

andreaskaris replied:
touché ;-)

dulek replied:
I meant no harm, it's a legitimate corner case. ;)

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+				return nil, nil, fmt.Errorf("requested IP address %s for node %s and port %s matches 2 different subnets, %s and %s",
 					ip, node.Name, serverPort.ID, matchingSubnet.ID, s.ID)
 			}
+
```
nit: blank line needed?

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+		return err
+	}
+
+	if matchingSubnet != nil {
```
findAssignSubetAndPort either returns nil,nil,error   or non-nil, non-nil, nil as far as I can see, so we can simplify this and remove the if and the return error in 5)?

andreaskaris reviewed (changes requested):
It'd be great to unit test `MovePrivateIP` in test function `TestOpenStackPlugin`

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
 	return fmt.Errorf("could not assign IP address %s to node %s", ip, node.Name)
 }
 
+func (o *OpenStack) MovePrivateIP(ip net.IP, nodeToAdd, nodeToDel *corev1.Node) error {
```
linter nit: openstack.go:303:1: exported method OpenStack.MovePrivateIP should have comment (...)

dulek commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+		}
+	}
+
+	// 5) The IP address does not fit in any of the attached networks' subnets.
```
True, thanks!

dulek commented on pkg/cloudprovider/openstack.go:
```diff
@@ -554,7 +617,38 @@ func (o *OpenStack) reserveNeutronIPAddress(s neutronsubnets.Subnet, ip net.IP,
 	p, err := neutronports.Create(o.neutronClient, opts).Extract()
+
 	if err != nil {
+		// Let's check if error suggests that port with that IP already exists in that subnet.
+		if _, ok := err.(*gophercloud.ErrDefault409); ok {
+			// If so, let's get the port and check if it's our reservation port. It's possible we've created it earlier.
+			// It's also possible user configured a taken range, we should erorr out in such case.
+			page, nestedErr := neutronports.List(o.neutronClient, neutronports.ListOpts{
+				NetworkID:   s.NetworkID,
+				FixedIPs:    []neutronports.FixedIPOpts{{SubnetID: s.ID, IPAddress: ip.String()}},
+				DeviceOwner: egressIPTag,
+				DeviceID:    generateDeviceID(serverID),
+			}).AllPages()
+			if nestedErr != nil {
+				return nil, nestedErr
+			}
+			ports, nestedErr := neutronports.ExtractPorts(page)
+			if nestedErr != nil {
+				return nil, nestedErr
+			}
+			if len(ports) > 1 {
+				// This is unexpected, Neutron should not allow multiple ports with the same IP
+				// in one subnet
+				return p, fmt.Errorf(
+					"bogus result from Neutron, multiple ports with IP %s in subnet %s", ip.String(), s.ID)
+			} else if len(ports) == 0 {
+				// Apparently it's not our port, return the original error.
+				return nil, err
+			}
+			// Otherwise we've got our port, let's return it.
+			return &ports[0], nil
+		}
+		// Regular error, just return it.
```
This is part I'll need to test - not sure about the error gophercloud returns in such case.

dulek commented on pkg/cloudprovider/openstack.go:
```diff
@@ -554,7 +617,38 @@ func (o *OpenStack) reserveNeutronIPAddress(s neutronsubnets.Subnet, ip net.IP,
+				DeviceOwner: egressIPTag,
+				DeviceID:    generateDeviceID(serverID),
+			}).AllPages()
+			if nestedErr != nil {
```
BTW - what's the correct pattern to nest errors? This one smells bad. I guess a separate function might look better?

andreaskaris replied:
you can wrap errors with %w:
https://rollbar.com/blog/golang-wrap-and-unwrap-error/#

There's a repo that implements this:
github.com/pkg/errors
With Wrap functions

You can join multiple errors with:
import (
apierrors "k8s.io/apimachinery/pkg/util/errors"
)
 apierrors.NewAggregate(errors)
https://pkg.go.dev/k8s.io/apimachinery/pkg/util/errors#NewAggregate

dulek replied:
I just changed the logic instead, I wasn't exactly wanting to wrap errors.

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,45 +219,126 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+// MovePrivateIP implements moving the IP from one node to another to serve cases like a failover.
+// What's different from calling ReleasePrivateIP followed by AssignPrivateIP is that the reservation
+// Neutron port will not get deleted - MovePrivateIP will only change the allowed_address_pairs on the node's
+// ports to remove IP from nodeToDel and add it to nodeToAdd.
```
Add something like: "Additionally, MovePrivateIP will recreate a missing reservation port (this is a corner case and should not happen in normal operation)"

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -554,7 +617,38 @@ func (o *OpenStack) reserveNeutronIPAddress(s neutronsubnets.Subnet, ip net.IP,
+		// Let's check if error suggests that port with that IP already exists in that subnet.
+		if _, ok := err.(*gophercloud.ErrDefault409); ok {
+			// If so, let's get the port and check if it's our reservation port. It's possible we've created it earlier.
+			// It's also possible user configured a taken range, we should erorr out in such case.
```
erorr -> error

andreaskaris commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,45 +219,126 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	}
+
+	// This call is to double-check if the reservation port exists, function
+	// will return okay if the port with that IP already exists in the subnet.
```
... will return okay if the port with that IP already exists in the subnet, otherwise it will create the missing reservation port.

kyrtapz commented on pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go:
```diff
@@ -179,24 +179,96 @@ func (c *CloudPrivateIPConfigController) SyncHandler(key string) error {
 
-	// At most one of nodeToAdd or nodeToDel will be set
-	nodeToAdd, nodeToDel := c.computeOp(cloudPrivateIPConfig)
+	nodeNameToAdd, nodeNameToDel := c.computeOp(cloudPrivateIPConfig)
```
NIT: For future reference, it is generally better to do variable renaming as a separate commit. It makes reviewing actual changes a lot easier.
You don't need to change it in this PR.

kyrtapz commented on pkg/cloudprovider/openstack.go:
```diff
@@ -237,45 +219,126 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+		return err
+	}
+
+	// 2) Reserve the IP address on the subnet by creating a new unattached neutron port.
```
Do the numbers in comments still make sense in `AssignPrivateIP` I don't think we need them with the introduction of `findAssignSubnetAndPort`

kyrtapz commented on pkg/cloudprovider/openstack.go:
```diff
@@ -554,7 +617,38 @@ func (o *OpenStack) reserveNeutronIPAddress(s neutronsubnets.Subnet, ip net.IP,
+				// Apparently it's not our port, return the original error.
+				return nil, err
+			}
+			// Otherwise we've got our port, let's return it.
```
can we add a info log saying we will reuse an existing port?

dulek commented:
I think I addressed all of the comments. I still have an issue with ovn-kubernetes, where CloudPrivateIPConfig is getting deleted after a successful move operation. I'm not sure if this isn't an ovn-kubernetes bug and I'll be digging that.

dulek commented:
I fixed ovn-kubernetes issue here: https://github.com/ovn-org/ovn-kubernetes/pull/3403. Please take a look.

dulek commented:
/retest

dulek commented:
@kyrtapz, @andreaskaris: https://github.com/ovn-org/ovn-kubernetes/pull/3403 is merged, so I think we can reconsider this one.

kyrtapz commented:
/lgtm
